pkg/kubelet/kuberuntime/kuberuntime_container.go

```diff
@@ -744,6 +744,18 @@ func findNextInitContainerToRun(pod *v1.Pod, podStatus *kubecontainer.PodStatus)
 		return nil, nil, true
 	}
 
+	// If any of the main containers have status and are Running, then all init containers must
+	// have been executed at some point in the past.  However, they could have been removed
+	// from the container runtime now, and if we proceed, it would appear as if they
+	// never ran and will re-execute improperly.
+	for i := range pod.Spec.Containers {
+		container := &pod.Spec.Containers[i]
+		status := podStatus.FindContainerStatusByName(container.Name)
```

Comment:
@sjenning If instead of `FindContainerStatusByName()` we use `podutil.GetContainerStatus()` (from k8s.io/kubernetes/pkg/api/v1/pod) we get this type that has more information instead: https://pkg.go.dev/k8s.io/api/core/v1#ContainerStatus. Seems confusing but is not really the same, IIRC.

So, if this loop instead of `pod.Spec.Containers` we iterate `pod.Spec.InitContainers` can't we get the status that it was already run? Or is that lost for some reason too in the bug you are chasing @sjenning?

IIUC this might be the case (untested, though) and it seems more robust to just see "have we run this? Then run". Otherwise if the gc happens while initContainers didn't finish, we might run some initContainers twice with this approach AFAIK. Also, this check (if it works as I hope :D) should work fine with the concern @derekwaynecarr had on [pod restart reasons](https://kubernetes.io/docs/concepts/workloads/pods/init-containers/#pod-restart-reasons)

Am I missing something? Probably I am :)

Reply:
Yes it is confusing.

In this code path, we are acting on the authoritative container status from the runtime i.e. `kubecontainer.Status`, not what the kubelet has previously reported to the apiserver i.e. `v1.ContainerStatus`, which is what `GetContainerStatus()` would give you.

Additionally, there are cases in which we do want the initContainers to re-run, so just because they have run in the past don't mean they shouldn't run again.  This is what Derek was saying here https://github.com/kubernetes/kubernetes/pull/96572#discussion_r523184267

That is why I expanded the main container check to not only check for status but also check that the main container is Running, since,  in all situations where the init containers need to be run again, the main containers are not running.

Reply:
Ohh, I didn't know we want to re-run initContainers in such a case, sorry! Then this makes total sense, that is what I was missing :)